vault/cors.go

```diff
@@ -0,0 +1,161 @@
+package vault
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/vault/helper/strutil"
+	"github.com/hashicorp/vault/logical"
+)
+
+var errCORSNotConfigured = errors.New("CORS is not configured")
+
+var preflightHeaders = map[string]string{
+	"Access-Control-Allow-Headers":     "*",
+	"Access-Control-Max-Age":           "1800",
+	"Access-Control-Allow-Credentials": "true",
+}
+
+var allowedMethods = []string{
+	http.MethodDelete,
+	http.MethodGet,
+	http.MethodOptions,
+	http.MethodPost,
+	http.MethodPut,
+	"LIST", // LIST is not an official HTTP method, but Vault supports it.
+}
+
+// CORSConfig stores the state of the CORS configuration.
+type CORSConfig struct {
+	Enabled        bool     `json:"enabled"`
+	AllowedOrigins []string `json:"allowed_origins"`
+	sync.RWMutex
+}
+
+func (c *Core) saveCORSConfig() error {
+	view := c.systemBarrierView.SubView("config/")
+
+	entry, err := logical.StorageEntryJSON("cors", c.corsConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create CORS confif entry: %v", err)
+	}
+
+	if err := view.Put(entry); err != nil {
+		return fmt.Errorf("failed to save CORS config: %v", err)
+	}
+
+	return nil
+}
+
+func (c *Core) loadCORSConfig() error {
+	view := c.systemBarrierView.SubView("config/")
+
+	// Load the config in
+	out, err := view.Get("cors")
+	if err != nil {
+		return fmt.Errorf("failed to read CORS config: %v", err)
+	}
+	if out == nil {
+		return nil
+	}
+
+	config := new(CORSConfig)
+	err = out.DecodeJSON(config)
+	if err != nil {
+		return err
+	}
+
+	c.corsConfig = config
+
+	return nil
+}
+
+// Enable takes either a '*' or a comma-seprated list of URLs that can make
+// cross-origin requests to Vault.
+func (c *CORSConfig) Enable(s string) error {
+	c.RWMutex.Lock()
+	defer c.RWMutex.Unlock()
+
+	if strings.Contains("*", s) && len(s) > 1 {
+		return errors.New("wildcard must be the only value")
+	}
+
+	c.AllowedOrigins = strings.Split(s, ",")
+	c.Enabled = true
+
+	return nil
+}
+
+// Get returns the state of the CORS configuration.
+func (c *CORSConfig) Get() *CORSConfig {
+	c.RWMutex.Lock()
```

Comment:
This is racy, because the locking you implemented is supposed to protect the values, not the struct itself, so if a caller gets the struct back and reads the values without then knowing that they need to lock it, you have a race condition.